mailbox/event/event-cassandra/src/main/java/org/apache/james/mailbox/events/CassandraEventDeadLettersDAO.java

```diff
@@ -0,0 +1,131 @@
+        this.executor = new CassandraAsyncExecutor(session);
+        this.eventSerializer = eventSerializer;
+        this.insertStatement = prepareInsertStatement(session);
+        this.deleteStatement = prepareDeleteStatement(session);
+        this.selectAllGroupStatement = prepareSelectAllGroupStatement(session);
+        this.selectEventStatement = prepareSelectEventStatement(session);
+        this.selectEventIdsWithGroupStatement = prepareSelectEventIdsWithGroupStatement(session);
+    }
+
+    private PreparedStatement prepareInsertStatement(Session session) {
+        return session.prepare(insertInto(TABLE_NAME)
+            .value(GROUP, bindMarker(GROUP))
+            .value(EVENT_ID, bindMarker(EVENT_ID))
+            .value(EVENT, bindMarker(EVENT)));
+    }
+
+    private PreparedStatement prepareDeleteStatement(Session session) {
+        return session.prepare(delete()
+            .from(TABLE_NAME)
+            .where(eq(GROUP, bindMarker(GROUP)))
+            .and(eq(EVENT_ID, bindMarker(EVENT_ID))));
+    }
+
+    private PreparedStatement prepareSelectAllGroupStatement(Session session) {
+        return session.prepare(select(GROUP)
+            .from(TABLE_NAME));
+    }
+
+    private PreparedStatement prepareSelectEventStatement(Session session) {
+        return session.prepare(select(EVENT)
+            .from(TABLE_NAME)
+            .where(eq(GROUP, bindMarker(GROUP)))
+            .and(eq(EVENT_ID, bindMarker(EVENT_ID))));
+    }
+
+    private PreparedStatement prepareSelectEventIdsWithGroupStatement(Session session) {
+        return session.prepare(select(EVENT_ID)
+            .from(TABLE_NAME)
+            .where(eq(GROUP, bindMarker(GROUP))));
+    }
+
+    Mono<Void> store(Group group, Event failedEvent) {
+        return executor.executeVoid(insertStatement.bind()
+            .setString(GROUP, group.asString())
+            .setUUID(EVENT_ID, failedEvent.getEventId().getId())
+            .setString(EVENT, eventSerializer.toJson(failedEvent)));
+    }
+
+    Mono<Void> removeEvent(Group group, Event.EventId failedEventId) {
+        return executor.executeVoid(deleteStatement.bind()
+            .setString(GROUP, group.asString())
+            .setUUID(EVENT_ID, failedEventId.getId()));
+    }
+
+    Mono<Event> retrieveFailedEvent(Group group, Event.EventId failedEventId) {
+        return executor.executeSingleRow(selectEventStatement.bind()
+            .setString(GROUP, group.asString())
+            .setUUID(EVENT_ID, failedEventId.getId()))
+            .map(row -> deserializeEvent(row.getString(EVENT)));
+    }
+
+    Flux<Event.EventId> retrieveEventIdsWithGroup(Group group) {
+        return executor.executeRows(selectEventIdsWithGroupStatement.bind()
+                .setString(GROUP, group.asString()))
+            .map(row -> Event.EventId.of(row.getUUID(EVENT_ID)));
+    }
+
+    Flux<Group> retrieveAllGroups() {
```

Comment:
While this solution "does the job" for an API perspective, full scan is clearly not performance optimal.

I propose you write in a separated PR another DAO for maintaining a `Set<Group>` table on Cassandra that we could be using here...

Reply:
PR: https://github.com/linagora/james-project/pull/2188 solved it.